subprojects/core/src/main/java/org/gradle/api/internal/plugins/ImperativeOnlyPluginTarget.java

```diff
@@ -40,7 +50,38 @@ public ConfigurationTargetIdentifier getConfigurationTargetIdentifier() {
     public void applyImperative(@Nullable String pluginId, Plugin<?> plugin) {
         // TODO validate that the plugin accepts this kind of argument
         Plugin<T> cast = uncheckedCast(plugin);
-        cast.apply(target);
+        try {
+            cast.apply(target);
+        } catch (ClassCastException e) {
+            maybeThrowOnTargetMismatch(plugin);
+            throw e;
+        }
+    }
+
+    private void maybeThrowOnTargetMismatch(Plugin<?> plugin) {
+        Type typeParameter = TypeUtils.getTypeArguments(plugin.getClass(), Plugin.class).get(Plugin.class.getTypeParameters()[0]);
+        if (!(typeParameter instanceof Class<?>)) {
+            return;
+        }
+
+        PluginTargetType actualTargetType = PluginTargetType.from((Class<?>) typeParameter);
+        if (actualTargetType == null || targetType.equals(actualTargetType)) {
```

Comment:
```suggestion
        if (actualTargetType == null || targetType.isAssignableFrom(actualTargetType)) {
```

Reply:
This check is between the enum values of type `PluginTargetType`, so equals should be correct here.

The place where we determine the plugin target type from the type parameter already uses `isAssignableFrom`:
https://github.com/gradle/gradle/pull/29721/files#diff-570a003c8603508ba2def36cf3b4e38d2c40aa03ea313bdb9ea4cd8dba332b75R48